File: app/src/main/java/com/example/music/data/LocalMusicLoader.java
package com.example.music.data;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import com.example.music.data.model.Album;
import com.example.music.data.model.Artist;
import com.example.music.data.model.Track;
import java.util.ArrayList;
import java.util.List;

public class LocalMusicLoader {

    public static List<Track> loadLocalTracks(Context context) {
        List<Track> tracks = new ArrayList<>();
        ContentResolver contentResolver = context.getContentResolver();
        Uri uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
        String selection = MediaStore.Audio.Media.IS_MUSIC + "!= 0";
        String sortOrder = MediaStore.Audio.Media.TITLE + " ASC";

        Cursor cursor = contentResolver.query(uri, null, selection, null, sortOrder);

        if (cursor != null && cursor.moveToFirst()) {
            int idColumn = cursor.getColumnIndex(MediaStore.Audio.Media._ID);
            int titleColumn = cursor.getColumnIndex(MediaStore.Audio.Media.TITLE);
            int artistColumn = cursor.getColumnIndex(MediaStore.Audio.Media.ARTIST);
            int albumColumn = cursor.getColumnIndex(MediaStore.Audio.Media.ALBUM);
            int albumIdColumn = cursor.getColumnIndex(MediaStore.Audio.Media.ALBUM_ID);

            do {
                long id = cursor.getLong(idColumn);
                String title = cursor.getString(titleColumn);
                String artistName = cursor.getString(artistColumn);
                String albumName = cursor.getString(albumColumn);
                long albumId = cursor.getLong(albumIdColumn);

                Uri trackUri = Uri.withAppendedPath(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, String.valueOf(id));
                Uri albumArtUri = Uri.parse("content://media/external/audio/albumart/" + albumId);

                Artist artist = new Artist();
                artist.setName(artistName);

                Album album = new Album();
                album.setTitle(albumName);
                album.setCoverMedium(albumArtUri.toString());
                album.setCoverBig(albumArtUri.toString());

                Track track = new Track();
                track.setId(id);
                track.setTitle(title);
                track.setArtist(artist);
                track.setAlbum(album);
                track.setPreview(trackUri.toString());

                tracks.add(track);
            } while (cursor.moveToNext());
            cursor.close();
        }

        return tracks;
    }
}

File: app/src/main/java/com/example/music/data/LocalMusicLoader.kt
package com.example.music.data

import android.content.ContentUris
import android.content.Context
import android.net.Uri
import android.provider.MediaStore
import com.example.music.data.model.Album
import com.example.music.data.model.Artist
import com.example.music.data.model.Track

object LocalMusicLoader {
    @JvmStatic
    fun loadLocalTracks(context: Context): List<Track> {
        val tracks = mutableListOf<Track>()
        val projection = arrayOf(
            MediaStore.Audio.Media._ID,
            MediaStore.Audio.Media.TITLE,
            MediaStore.Audio.Media.ARTIST,
            MediaStore.Audio.Media.ALBUM,
            MediaStore.Audio.Media.ALBUM_ID,
            MediaStore.Audio.Media.DURATION
        )
        val selection = "${MediaStore.Audio.Media.IS_MUSIC} != 0"
        val sortOrder = "${MediaStore.Audio.Media.TITLE} ASC"
        val contentResolver = context.contentResolver
        val uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI

        contentResolver.query(uri, projection, selection, null, sortOrder)?.use { cursor ->
            val idColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media._ID)
            val titleColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.TITLE)
            val artistColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST)
            val albumColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM)
            val albumIdColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM_ID)
            val durationColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DURATION)

            while (cursor.moveToNext()) {
                val id = cursor.getLong(idColumn)
                val title = cursor.getString(titleColumn)
                val artistName = cursor.getString(artistColumn)
                val albumName = cursor.getString(albumColumn)
                val albumId = cursor.getLong(albumIdColumn)
                val duration = cursor.getInt(durationColumn)

                val trackUri: Uri = ContentUris.withAppendedId(
                    MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                    id
                )
                val albumArtUri: Uri = ContentUris.withAppendedId(
                    Uri.parse("content://media/external/audio/albumart"),
                    albumId
                )

                tracks += Track(
                    id = id,
                    title = title,
                    artist = Artist(0, artistName ?: "Unknown Artist"),
                    album = Album(
                        id = albumId,
                        title = albumName ?: "Unknown Album",
                        coverMedium = albumArtUri.toString(),
                        coverBig = albumArtUri.toString()
                    ),
                    preview = trackUri.toString(),
                    duration = duration
                )
            }
        }
        return tracks
    }
}
